src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,155 @@
+            previousNode.next = addingNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T addedNode : list) {
+            add(addedNode);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, "get");
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, "set");
+        Node<T> nodeToSet = findNodeByIndex(index);
+        T oldValue = nodeToSet.value;
+        nodeToSet.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, "remove");
+        if (isEmpty()) {
+            return null;
+        }
+        Node<T> nodeToRemove = findNodeByIndex(index);
+        unLink(nodeToRemove);
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        if (isEmpty()) {
+            return false;
+        }
+        Node<T> nodeToRemove;
+        for (nodeToRemove = head; nodeToRemove != null; nodeToRemove = nodeToRemove.next) {
+            if (object == nodeToRemove.value
+                    || object != null && object.equals(nodeToRemove.value)) {
+                unLink(nodeToRemove);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unLink(Node<T> node) {
+        Node<T> previousNode;
+        Node<T> nextNode;
+        if (node.prev == null && node.next == null) {
+            size--;
+        } else if (node.prev == null) {
+            head = node.next;
+            head.prev = null;
+            size--;
+        } else if (node.next == null) {
+            tail = node.prev;
+            tail.next = null;
+            size--;
+        } else {
+            previousNode = node.prev;
+            nextNode = node.next;
+            previousNode.next = nextNode;
+            nextNode.prev = previousNode;
+            size--;
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> soughtNode;
+        if (index < size / 2) {
+            soughtNode = head;
+            for (int i = 0; i < index; i++) {
+                soughtNode = soughtNode.next;
+            }
+        } else {
+            soughtNode = tail;
+            for (int i = size - 1; i > index; --i) {
+                soughtNode = soughtNode.prev;
+            }
+        }
+        return soughtNode;
+    }
+
+    private void checkIndex(int index, String operationType) throws IndexOutOfBoundsException {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Can't " + operationType
```

Comment:
better pass index to the exception instead of operationType, because operationType we can get from stacktrace